go/libkb/upak_lite.go

```diff
@@ -0,0 +1,273 @@
+	return
+}
+
+func (l *HighSigChainLoader) LoadFromServer() (err error) {
+	srv := l.GetMerkleTriple()
+	l.dirtyTail, err = l.chain.LoadFromServer(l.M(), srv, l.selfUID())
+	return
+}
+
+func (hsc *HighSigChain) LoadFromServer(m MetaContext, t *MerkleTriple, selfUID keybase1.UID) (dirtyTail *MerkleTriple, err error) {
+	// get the high sigs from the server
+	// ------------------
+	m, tbs := m.WithTimeBuckets()
+
+	apiArg := APIArg{
+		Endpoint:    "sig/get_high",
+		SessionType: APISessionTypeREQUIRED,
+		Args:        HTTPArgs{"uid": S{Val: hsc.uid.String()}},
+		MetaContext: m,
+	}
+	resp, finisher, err := m.G().API.GetResp(apiArg)
+	if err != nil {
+		return nil, err
+	}
+	if finisher != nil {
+		defer finisher()
+	}
+	recordFin := tbs.Record("HighSigChain.LoadFromServer.ReadAll")
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		recordFin()
+		return nil, err
+	}
+	recordFin()
+
+	// parse the response
+	// ------------------
+	if val, err := jsonparser.GetInt(body, "status", "code"); err == nil {
+		if keybase1.StatusCode(val) == keybase1.StatusCode_SCDeleted {
+			return nil, UserDeletedError{}
+		}
+	}
+	var links ChainLinks
+	var lastLink *ChainLink
+
+	jsonparser.ArrayEach(body, func(value []byte, dataType jsonparser.ValueType, offset int, inErr error) {
+		var link *ChainLink
+
+		parentSigChain := &SigChain{} // because we don't want the cache to use these
+		link, err = ImportLinkFromServer(m.G(), parentSigChain, value, selfUID)
+		links = append(links, link)
+		lastLink = link
+	}, "sigs")
+	foundTail, err := lastLink.checkAgainstMerkleTree(t)
+	if err != nil {
+		return nil, err
+	}
+	if !foundTail {
+		err = fmt.Errorf("Last link is not the tail")
+		return nil, err
+	}
+	dirtyTail = lastLink.ToMerkleTriple()
+
+	hsc.chainLinks = links
+	return dirtyTail, err
+}
+
+func (hsc *HighSigChain) VerifyChain(m MetaContext) (err error) {
+	// for each link, call VerifyLink on it
+	// also check the hprevs and seqnos and link.CheckNameAndID
+	// set that it's highChainVerified
+	for i := len(hsc.chainLinks) - 1; i >= 0; i-- {
+		curr := hsc.chainLinks[i]
+		if err = curr.VerifyLink(); err != nil {
+			return err
+		}
+		if i > 0 {
+			prev := hsc.chainLinks[i-1]
+			var expectedPrevID LinkID
+			var expectedPrevSeqno keybase1.Seqno
+			if curr.GetHighSkip() != nil {
+				expectedPrevSeqno = curr.GetHighSkip().Seqno
+				expectedPrevID = curr.GetHighSkip().Hash
+			} else {
+				// fallback to normal prevs if the link doesn't have a high_skip
+				expectedPrevSeqno = curr.GetSeqno() - 1
+				expectedPrevID = curr.GetPrev()
+			}
+			if !prev.id.Eq(expectedPrevID) {
+				return ChainLinkPrevHashMismatchError{fmt.Sprintf("Chain mismatch at seqno=%d", curr.GetSeqno())}
+			}
+			if prev.GetSeqno() != expectedPrevSeqno {
+				return ChainLinkWrongSeqnoError{fmt.Sprintf("Chain seqno mismatch at seqno=%d (previous=%d)", curr.GetSeqno(), prev.GetSeqno())}
+			}
```

Comment:
could also check for 0 seqno/null id for the first link

Reply:
this is done.